### PR TITLE
Print start of testhost standard error message

### LINKED ISSUE #1688: Debug.Assert(false) causes poor experience in tests
## Description
When a Debug.Assert fails during unit tests, there is no way to know what failed from looking at the output.

I've looked at https://github.com/Microsoft/vstest/issues/1022, but the behavior there seems to have regressed in the latest builds. I am not getting an output that says an assert failed, and in no place does it say a stack trace of the failing assert.

## Steps to reproduce
Using this project
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp2.1</TargetFramework>
    <IsPackable>false</IsPackable>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="15.8.0-preview-20180610-02" />
    <PackageReference Include="xunit" Version="2.3.1" />
    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
  </ItemGroup>
</Project>
```
and this code:
```c#
using System.Diagnostics;
using Xunit;

namespace XUnitTestProject1
{
    public class UnitTest1
    {
        [Fact]
        public void Test1()
        {
            Debug.Assert(false);
        }
    }
}
```

run the test in both `dotnet test` and in VS Test Explorer (15.8.0 Preview 4.0 [27907.0.d15.8])

## Expected behavior
I expect to at least get an error indicating a debug.assert failed and a stack trace of where in my code the assert failed.

## Actual behavior
Here is the full output from the command line.  I get similar output in the VS Output window.  Neither place says that it is `UnitTest1` that is failing causing an assert. My only recourse is to run the test again and debug.

```
C:\users\eerhardt\source\XUnitTestProject1> dotnet test
Build started, please wait...
Build completed.

Test run for C:\users\eerhardt\source\XUnitTestProject1\bin\Debug\netcoreapp2.1\XUnitTestProject1.dll(.NETCoreApp,Version=v2.1)
Microsoft (R) Test Execution Command Line Tool Version 15.8.0-preview-20180605-02
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
The active test run was aborted. Reason: (Func`1 code)
   at Xunit.Sdk.TestRunner`1.RunAsync()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Xunit.Sdk.TestRunner`1.RunAsync()
   at Xunit.Sdk.TestCaseRunner`1.RunAsync() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\Runners\TestCaseRunner.cs:line 82
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Xunit.Sdk.TestCaseRunner`1.RunAsync()
   at Xunit.Sdk.XunitTestMethodRunner.RunTestCaseAsync(IXunitTestCase testCase) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\Runners\XunitTestMethodRunner.cs:line 45
   at Xunit.Sdk.TestMethodRunner`1.RunTestCasesAsync() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\Runners\TestMethodRunner.cs:line 136
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Xunit.Sdk.TestMethodRunner`1.RunTestCasesAsync()
   at Xunit.Sdk.TestMethodRunner`1.RunAsync() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\Runners\TestMethodRunner.cs:line 106
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Xunit.Sdk.TestMethodRunner`1.RunAsync()
   at Xunit.Sdk.TestClassRunner`1.RunTestMethodsAsync() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\Runners\TestClassRunner.cs:line 213
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Xunit.Sdk.TestClassRunner`1.RunTestMethodsAsync()
   at Xunit.Sdk.TestClassRunner`1.RunAsync() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\Runners\TestClassRunner.cs:line 171
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Xunit.Sdk.TestClassRunner`1.RunAsync()
   at Xunit.Sdk.TestCollectionRunner`1.RunTestClassesAsync() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\Runners\TestCollectionRunner.cs:line 130
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Xunit.Sdk.TestCollectionRunner`1.RunTestClassesAsync()
   at Xunit.Sdk.TestCollectionRunner`1.RunAsync() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\Runners\TestCollectionRunner.cs:line 101
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Xunit.Sdk.TestCollectionRunner`1.RunAsync()
   at Xunit.Sdk.XunitTestAssemblyRunner.<>c__DisplayClass14_2.<RunTestCollectionsAsync>b__2() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\Runners\XunitTestAssemblyRunner.cs:line 184
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
   at System.Threading.Tasks.Task.ExecuteEntry()
   at System.Threading.Tasks.SynchronizationContextTaskScheduler.<>c.<.cctor>b__8_0(Object s)
   at Xunit.Sdk.MaxConcurrencySyncContext.RunOnSyncContext(SendOrPostCallback callback, Object state) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\MaxConcurrencySyncContext.cs:line 107
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at Xunit.Sdk.MaxConcurrencySyncContext.WorkerThreadProc() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\MaxConcurrencySyncContext.cs:line 89
   at Xunit.Sdk.XunitWorkerThread.<>c.<QueueUserWorkItem>b__5_0(Object _) in C:\Dev\xunit\xunit\src\common\XunitWorkerThread.cs:line 37
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)


Test Run Aborted.
```

## Diagnostic logs
[log.txt](https://github.com/Microsoft/vstest/files/2193389/log.txt)

## Environment
```
> dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   2.1.400-preview-009063
 Commit:    dd0179a67c

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17134
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.1.400-preview-009063\

Host (useful for support):
  Version: 2.1.1
  Commit:  6985b9f684

.NET Core SDKs installed:
  1.1.9 [C:\Program Files\dotnet\sdk]
  2.1.1 [C:\Program Files\dotnet\sdk]
  2.1.2 [C:\Program Files\dotnet\sdk]
  2.1.4 [C:\Program Files\dotnet\sdk]
  2.1.100 [C:\Program Files\dotnet\sdk]
  2.1.200 [C:\Program Files\dotnet\sdk]
  2.1.201 [C:\Program Files\dotnet\sdk]
  2.1.300 [C:\Program Files\dotnet\sdk]
  2.1.400-preview-009063 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 1.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
```

/cc @singhsarab @smadala

### PR DESCRIPTION
**Pull request by smadala**
## Description
- Prints start of stderr stacktrace
- Increase max length of stderr message.

## Related issue
Fixes #1688

### CODE REVIEW

File: src/Microsoft.TestPlatform.CoreUtilities/Utilities/StringExtensions.cs [hunk 1]
```diff
@@ -17,5 +19,28 @@ public static string AddDoubleQuote(this string value)
         {
             return "\"" + value + "\"";
         }
+
+        public static void AppendToStringBuilderBasedOnMaxLength(this string data, StringBuilder result)
```
Comment (abhishkk) [inline src/Microsoft.TestPlatform.CoreUtilities/Utilities/StringExtensions.cs]: Doesn't look like string extension. May be a string builder extension.
`public static void AppendSafe(this StringBuilder stringBuilder, string data)`

Comment (abhishkk) [inline src/Microsoft.TestPlatform.CoreUtilities/Utilities/StringExtensions.cs]: Can we write comment as well for this method?

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DataCollectionLauncher.cs [hunk 2]
```diff
@@ -36,17 +36,12 @@ public DataCollectionLauncher(IProcessHelper processHelper, IMessageLogger messa
         {
             this.processHelper = processHelper;
             this.messageLogger = messageLogger;
-            this.processStdError = new StringBuilder(this.ErrorLength, this.ErrorLength);
+            this.processStdError = new StringBuilder(0, CoreUtilities.Constants.StandardErrorMaxLength);
```
Comment (abhishkk) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DataCollectionLauncher.cs]: When setting capacity to zero, will stringBuilder not be perf inefficient? As it needs to keep on copying and pasting the content to new instance whenever grow is required.

File: test/Microsoft.TestPlatform.TestHostProvider.UnitTests/Hosting/TestHostManagerCallbacksTests.cs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        {
+            this.testHostProcessStdError = new StringBuilder(0, 5);
+            this.testHostProcessStdError.Append("1234");
+            TestHostManagerCallbacks.ErrorReceivedCallback(this.testHostProcessStdError, "5678");
+
+            Assert.AreEqual("12345", this.testHostProcessStdError.ToString());
+        }
+
+        [TestMethod]
+        public void ErrorReceivedCallbackShouldAppendEntireStringEvenItReachesToMaxLength()
```
Comment (abhishkk) [inline test/Microsoft.TestPlatform.TestHostProvider.UnitTests/Hosting/TestHostManagerCallbacksTests.cs]: `EvenItReachesToMaxLength`? In code, max length is not reached.

File: src/Microsoft.TestPlatform.CoreUtilities/Utilities/StringExtensions.cs [hunk 4]
```diff
@@ -17,5 +19,28 @@ public static string AddDoubleQuote(this string value)
...
+            if (!string.IsNullOrEmpty(data))
+            {
+                // Don't append more data if already reached max length.
+                if (result.Length >= result.MaxCapacity)
+                {
+                    return;
+                }
+
+                // Add newline for readbility.
+                data += Environment.NewLine;
```
Comment (abhishkk) [inline src/Microsoft.TestPlatform.CoreUtilities/Utilities/StringExtensions.cs]: If we are writing stringBuilder extension, then it should not be job of stringBuilder to add a new line. It should be the user's responsibility to add a new line.

**Review by abhishkk — APPROVED** [review]
approved with suggestions.

File: src/Microsoft.TestPlatform.CoreUtilities/Constants.cs [hunk 5]
```diff
@@ -22,5 +22,11 @@ public class Constants
         /// Datacollector process name, without file extension(.exe/.dll)
         /// </summary>
         public const string DatacollectorProcessName = "datacollector";
+
+        /// <summary>
+        /// Number of character should be logged on child process exited with
+        /// error message on stadard error.
+        /// </summary>
+        public const int StandardErrorMaxLength = 20480; // 20 KB
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.CoreUtilities/Constants.cs]: I don't see a reason for increasing it by 5 times. 4096 was a reasonable buffer, we can probably make ti 8192 (2x), if user wants more info, they can always go ahead check the logs.

File: src/Microsoft.TestPlatform.CoreUtilities/Constants.cs [hunk 6]
```diff
@@ -22,5 +22,11 @@ public class Constants
         /// Datacollector process name, without file extension(.exe/.dll)
         /// </summary>
         public const string DatacollectorProcessName = "datacollector";
+
+        /// <summary>
+        /// Number of character should be logged on child process exited with
+        /// error message on stadard error.
```
[the remark this answers is not in the document]
Comment (smadala) [reply src/Microsoft.TestPlatform.CoreUtilities/Constants.cs]: Done.

File: src/Microsoft.TestPlatform.CoreUtilities/Constants.cs (hunk 5 above)
Comment (smadala) [reply src/Microsoft.TestPlatform.CoreUtilities/Constants.cs]: Made it 8192

File: src/Microsoft.TestPlatform.CoreUtilities/Utilities/StringExtensions.cs (hunk 1 above)
Comment (smadala) [reply src/Microsoft.TestPlatform.CoreUtilities/Utilities/StringExtensions.cs]: Create `StringBuilderExtensions` class with `AppendSafeWithNewLine()`.

Comment (smadala) [reply src/Microsoft.TestPlatform.CoreUtilities/Utilities/StringExtensions.cs]: Done.

File: src/Microsoft.TestPlatform.CoreUtilities/Utilities/StringExtensions.cs (hunk 4 above)
Comment (smadala) [reply src/Microsoft.TestPlatform.CoreUtilities/Utilities/StringExtensions.cs]: Create `StringBuilderExtensions` class with `AppendSafeWithNewLine()`.

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DataCollectionLauncher.cs (hunk 2 above)
Comment (smadala) [reply src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DataCollectionLauncher.cs]: We don't expect any stderr most of time. StringBuilder increase size exponentially.

File: test/Microsoft.TestPlatform.TestHostProvider.UnitTests/Hosting/TestHostManagerCallbacksTests.cs (hunk 3 above)
Comment (smadala) [reply test/Microsoft.TestPlatform.TestHostProvider.UnitTests/Hosting/TestHostManagerCallbacksTests.cs]: In Windows `NewLine` is two characters.

File: src/Microsoft.TestPlatform.CoreUtilities/Utilities/StringBuilderExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+    public static class StringBuilderExtensions
+    {
+        /// <summary>
+        /// Add double quote around string. Useful in case of path which has white space in between.
+        /// </summary>
+        /// <param name="result">string builder</param>
+        /// /// <param name="data">data to be appended.</param>
+        /// <returns></returns>
+
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.CoreUtilities/Utilities/StringBuilderExtensions.cs]: Nit:Extra line
Nit:readability* inside the method below

File: src/Microsoft.TestPlatform.CoreUtilities/Utilities/StringBuilderExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+namespace Microsoft.VisualStudio.TestPlatform.CoreUtilities.Extensions
+{
+    using System;
+    using System.Text;
+
+    public static class StringBuilderExtensions
+    {
+        /// <summary>
+        /// Add double quote around string. Useful in case of path which has white space in between.
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.CoreUtilities/Utilities/StringBuilderExtensions.cs]: Summary needs to be corrected.

File: src/Microsoft.TestPlatform.CoreUtilities/Utilities/StringExtensions.cs [hunk 9]
```diff
@@ -17,5 +17,28 @@ public static string AddDoubleQuote(this string value)
         {
             return "\"" + value + "\"";
         }
+
+        public static void AppendToStringBuilderBasedOnMaxLength(this string data, StringBuilder result)
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.CoreUtilities/Utilities/StringExtensions.cs]: Still see this code, please remove this.

File: src/Microsoft.TestPlatform.CoreUtilities/Utilities/StringBuilderExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+                    return;
+                }
+
+                // Add newline for readbility.
+                data += Environment.NewLine;
+
+                // Append sub string of data if appending all the data exceeds max capacity.
+                if (result.Length + data.Length >= result.MaxCapacity)
+                {
+                    data = data.Substring(0, result.MaxCapacity - result.Length);
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CoreUtilities/Utilities/StringBuilderExtensions.cs]: What's the idea here, are we saying the original Sting content will remain, & only add small part of new data?

File: src/Microsoft.TestPlatform.CoreUtilities/Utilities/StringExtensions.cs (hunk 9 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CoreUtilities/Utilities/StringExtensions.cs]: Why is this method needed?

File: src/Microsoft.TestPlatform.CoreUtilities/Utilities/StringBuilderExtensions.cs (hunk 10 above)
Comment (smadala) [reply src/Microsoft.TestPlatform.CoreUtilities/Utilities/StringBuilderExtensions.cs]: Yes.

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DataCollectionLauncher.cs [hunk 11]
```diff
@@ -6,7 +6,7 @@ namespace Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.DataCollection
     using System;
     using System.Collections.Generic;
     using System.Text;
-
+    using CoreUtilities.Extensions;
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DataCollectionLauncher.cs]: Nit: Please use full namespace

File: src/Microsoft.TestPlatform.CoreUtilities/Utilities/StringExtensions.cs (hunk 9 above)
Comment (smadala) [reply src/Microsoft.TestPlatform.CoreUtilities/Utilities/StringExtensions.cs]: This is removed in latest commits.

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DataCollectionLauncher.cs (hunk 2 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DataCollectionLauncher.cs]: Should we initialize it with something other than that of 0 length?

**Pull request merged**